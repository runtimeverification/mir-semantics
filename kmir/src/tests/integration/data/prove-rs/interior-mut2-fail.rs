use std::cell::Cell;

struct Counter {
    value: Cell<i32>,
}

impl Counter {
    fn new(start: i32) -> Self {
        Counter { value: Cell::new(start) }
    }

    fn increment(&self) {
        self.value.set(self.get() + 1);
    }

    fn get(&self) -> i32 {
        self.value.get()
    }
}

fn main() {
    let counter = Counter::new(0);
    // println!("Before: {}", counter.get());

    // We only have &counter, but can still mutate inside
    counter.increment();
    counter.increment();

    assert!(2 == counter.get());
    // println!("After: {}", counter.get());
}
