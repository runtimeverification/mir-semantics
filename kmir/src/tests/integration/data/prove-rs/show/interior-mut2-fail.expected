
┌─ 1 (root, init)
│   #init ( symbol ( "interior_mut2_fail" ) globalAllocEntry ( 1 , Memory ( allocati
│   function: main
│   span: 120
│
│  (151 steps)
└─ 3 (stuck, leaf)
    #readProjection ( typedValue ( thunk ( #cast ( typedValue ( thunk ( rvalueAddres
    span: 53


┌─ 2 (root, leaf, target, terminal)
│   #EndProgram


